src/main/java/jcifs/smb/SmbEnumerationUtil.java

```diff
@@ -0,0 +1,364 @@
+
+        if ( locator.getType() != SmbConstants.TYPE_SERVER )
+            throw new SmbException("The requested list operations is invalid: " + u.toString());
+
+        Set<FileEntry> set = new HashSet<>();
+
+        if ( tc.getDfs().isTrustedDomain(tc, locator.getServer()) ) {
+            /*
+             * The server name is actually the name of a trusted
+             * domain. Add DFS roots to the list.
+             */
+            try {
+                entries = doDfsRootEnum(tc, locator, locator.getAddress());
+                for ( int ei = 0; ei < entries.length; ei++ ) {
+                    FileEntry e = entries[ ei ];
+                    if ( !set.contains(e) && ( fnf == null || fnf.accept(parent, e.getName()) ) ) {
+                        set.add(e);
+                    }
+                }
+            }
+            catch ( IOException ioe ) {
+                log.debug("DS enumeration failed", ioe);
+            }
+        }
+
+        SmbTreeConnection treeConn = SmbTreeConnection.create(tc);
+        try ( SmbTreeHandleImpl th = treeConn.connectHost(locator, locator.getServerWithDfs());
+              SmbSessionImpl session = th.getSession();
+              SmbTransportImpl transport = session.getTransport() ) {
+            try {
+                entries = doMsrpcShareEnum(tc, locator, transport.getRemoteAddress());
+            }
+            catch ( IOException ioe ) {
+                if ( th.isSMB2() ) {
+                    throw ioe;
+                }
+                log.debug("doMsrpcShareEnum failed", ioe);
+                entries = doNetShareEnum(th);
+            }
+            for ( int ei = 0; ei < entries.length; ei++ ) {
+                FileEntry e = entries[ ei ];
+                if ( !set.contains(e) && ( fnf == null || fnf.accept(parent, e.getName()) ) ) {
+                    set.add(e);
+                }
+            }
+
+        }
+        catch ( SmbException e ) {
+            throw e;
+        }
+        catch ( IOException ioe ) {
+            log.debug("doNetShareEnum failed", ioe);
+            throw new SmbException(u.toString(), ioe);
+        }
+        return new ShareEnumIterator(parent, set.iterator(), ff);
+    }
+
+
+    @SuppressWarnings ( "resource" )
+    static CloseableIterator<SmbResource> doEnum ( SmbFile parent, String wildcard, int searchAttributes, ResourceNameFilter fnf, ResourceFilter ff )
+            throws CIFSException {
+        DosFileFilter dff = unwrapDOSFilter(ff);
+        if ( dff != null ) {
+            if ( dff.wildcard != null )
+                wildcard = dff.wildcard;
+            searchAttributes = dff.attributes;
+        }
+        SmbResourceLocator locator = parent.getLocator();
+        if ( locator.getURL().getHost().isEmpty() ) {
+            // smb:// -> enumerate servers through browsing
+            Address addr;
+            try {
+                addr = locator.getAddress();
+            }
+            catch ( CIFSException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)